[RDY] vim-patch:8.1.0544: setting 'filetype' in a modeline causes an error

abdelhakeem opened the pull request:
Problem:    Setting 'filetype' in a modeline causes an error (Hirohito Higashi).
Solution:   Don't add the P_INSECURE flag when setting 'filetype' from a modeline.  Also for 'syntax'.
https://github.com/vim/vim/commit/916a818cea5ba05a5f2117407674461b8bee6832

fixes https://github.com/neovim/neovim/issues/9791

abdelhakeem committed:
vim-patch:8.1.0544: setting 'filetype' in a modeline causes an error

Problem:    Setting 'filetype' in a modeline causes an error (Hirohito
            Higashi).
Solution:   Don't add the P_INSECURE flag when setting 'filetype' from a
            modeline.  Also for 'syntax'.
https://github.com/vim/vim/commit/916a818cea5ba05a5f2117407674461b8bee6832

mhinz commented on src/nvim/option.c:
```diff
@@ -1359,6 +1359,10 @@ do_set (
           errmsg = e_trailing;
       } else {
+
+        int value_is_replaced = !prepending && !adding && !removing;
```
In the original patch, this line already exists. Thus this patch is based on another earlier patch that should be merged first.

mhinz replied:
https://github.com/vim/vim/commit/247bb7e43b47eb8407a1111bed60b61aceda52ad

abdelhakeem replied:
Shall I include it in a separate request or just merge it here?

mhinz replied:
Here, please.

abdelhakeem replied:
I found out we'll need a bunch of other patches before this one as well, some of which are not directly related to this issue, so I'm temporarily keeping this one aside and starting with the older ones...

mhinz replied:
Yeah, it's hard to keep up with porting patches while developing Nvim at the same time. So, any help there is more than welcome.

Maybe you used it for this commit already, but in case you didn't read it so far: https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-vim

abdelhakeem committed:
vim-patch:8.1.0540: may evaluate insecure value when appending to option

Problem:    May evaluate insecure value when appending to option.
Solution:   Set the secure flag when changing an option that was previously
            set insecurely.  Also allow numbers for the characters from
            'spelllang' that are used for LANG.vim.
https://github.com/vim/vim/commit/247bb7e43b47eb8407a1111bed60b61aceda52ad

abdelhakeem committed:
vim-patch:8.1.0539: cannot build without the sandbox

Problem:    Cannot build without the sandbox.
Solution:   Set the secure option instead of using the sandbox.  Also restrict
            the characters from 'spelllang' that are used for LANG.vim.
            (suggested by Yasuhiro Matsumoto)
https://github.com/vim/vim/commit/82e8c92ebef5afcac0c0fdb706ff163f9b3366f7

justinmk commented:
other than lint, is this ready? Or still WIP?

abdelhakeem commented:
> other than lint, is this ready? Or still WIP?

It's ready, planned to update it after fixing the linting errors

justinmk commented:
> Edit: do we actually bother about them? :) I can see the style it's yelling about being used kinda consistently throughout the code so I think this is ready now yeah

Yes, otherwise CI would always fail. The linter checks nearby lines so that old/legacy code is cleaned up gradually.

janlazo commented:
Need  vim-patch:8.1.1046 as noted in https://github.com/neovim/neovim/pull/9776#issuecomment-475961899.

abdelhakeem committed:
vim-patch:8.1.0544 followup: fix linting errors

abdelhakeem committed:
vim-patch:8.1.0544 followup: use vim_snprintf

blueyed commented:
Looks like there is too much linting - only code around the patch should be linted AFAIK.

abdelhakeem committed:
vim-patch:8.1.0544 followup: fix undeclared id for win builds

abdelhakeem commented:
> Looks like there is too much linting - only code around the patch should be linted AFAIK.

Should be ready soon, hopefully. Linting is done, just was fixing an issue with MSVC_32 and MINGW_64 builds (ironically introduced while fixing linting issues, my bad) and it's right now pending on AppVeyor.
Sorry for that, my first PR anyway uhum

blueyed commented:
What I've meant is that you should only fix linting for the code near/around the patch, not the whole file(s) (i.e. only the instances that CI complains about).

justinmk commented:
> What I've meant is that you should only fix linting for the code near/around the patch, not the whole file(s) (i.e. only the instances that CI complains about).

abdelhakeem commented:
> > What I've meant is that you should only fix linting for the code near/around the patch, not the whole file(s) (i.e. only the instances that CI complains about).
> 
> @abdelhakeem since you already did the work, just keep that big "lint" commit. It looks easy enough to review.

Yeah I'm keeping it.
Btw aren't the CI lint tests the same as those of `make lint`?

mhinz commented:
Personally I'm not a fan of these big lint commits as it makes normal `git blame` usage useless. The problem with such style commits is that afterwards mostly all lines will be touched by a single commit and you can't just `:Gblame` anymore to see _why_ or _when_ certain lines were changed.

And the followup patches can all be "added" to the original patch via `git commit --fixup <original patch SHA>`. Then either squash them locally (via `git rebase -i`) or push the fixup commits to GitHub as well and we later merge the PR via the `Sqush and merge` button.

justinmk commented on src/nvim/option.c:
```diff
@@ -1778,12 +1780,37 @@ do_set (
-            errmsg = did_set_string_option(opt_idx, (char_u **)varp,
-                new_value_alloced, oldval, errbuf, opt_flags);
+            {
+              unsigned int *p = insecure_flag(opt_idx, opt_flags);
```
`insecure_flag ` returns `uint32_t*` so that's what we would declare here.

justinmk commented on src/nvim/option.c:
```diff
@@ -1810,8 +1837,7 @@ do_set (
         if (opt_idx >= 0)
-          did_set_option(opt_idx, opt_flags,
-              !prepending && !adding && !removing);
+          did_set_option(opt_idx, opt_flags, value_is_replaced);
```
This commit is missing a `ASCII_ISALNUM` change. I'll fix this in the merge.

justinmk commented on src/nvim/option.c:
```diff
@@ -1778,12 +1780,37 @@ do_set (
+              // set with the P_INSECURE flag and is not
+              // completely replaced.
+              if (secure
+#ifdef HAVE_SANDBOX
```
compile-time feature-checks for `HAVE_XX` are never needed in Nvim, these can be left out.

Note that "features" (Vim features) are different than "platform capabilities". We do need to check platform stuff sometimes.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -117,7 +117,7 @@ read_buffer(
   if (retval == OK) {
     // Delete the binary lines.
-    while (--line_count >= 0) {
+    while (line_count-- > 0) {
```
This is a case we would not change, and may cause a bug. 

pre/post-decrement have different behavior in expressions (as opposed to statements). The linter only warns about about _statements_, not expressions.

justinmk commented on src/nvim/option.c:
```diff
@@ -1958,8 +1958,8 @@ static char_u *illegal_char(char_u *errbuf, size_t errbuflen, int c)
-  snprintf((char *)errbuf, errbuflen, _("E539: Illegal character <%s>"),
-           (char *)transchar(c));
+  vim_snprintf((char *)errbuf, errbuflen, _("E539: Illegal character <%s>"),
+               (char *)transchar(c));
```
what was the reason for this change?

abdelhakeem replied:
I would use `snprintf` but used this for consistency since I found it already being used like here:
https://github.com/neovim/neovim/blob/dfb7f6b3492edfd60583f76ee064ac76b1b913d8/src/nvim/option.c#L4405-L4406

justinmk commented on src/nvim/buffer.c:
```diff
@@ -4004,7 +4122,8 @@ int build_stl_str_hl(
         *t = t[-3];
-        *++t = 0;
+        t++;
+        *t = 0;
```
Similar case, we would not change this. And linter should not warn about it.

justinmk commented on src/nvim/buffer.c:
```diff
@@ -3942,8 +4059,9 @@ int build_stl_str_hl(
-      if (out_p + 20 > out_end_p)
-        break;                  /* not sufficient space */
+      if (out_p + 20 > out_end_p) {
+        break;                  //  not sufficient space
```
Only one space should be after `//`, not two. I mention this for future reference. I will fix it in this PR.
